runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java

```diff
@@ -330,35 +332,48 @@ public LayoutBuilder mergePreviousSegment(int segmentIndex) {
      * Moves a responsive server to the top of the sequencer server list.
      * If all have failed, throws exception.
      *
-     * @param endpoints a non null set of Strings representing failed endpoints
-     *
+     * @param excludedNodes a non null set of Strings representing excluded endpoints
      * @return this builder
-     *
-     * @throws LayoutModificationException is thrown if none of the sequencers in layout
-     *         can be moved to the top of sequencer server list due to being unresponsive
-     *         or a failed endpoint.
+     * @throws LayoutModificationException if none of the sequencers in the layout can be
+     *                                     moved to the top of sequencer server list due
+     *                                     to being unresponsive or a failed endpoint.
      */
-    public LayoutBuilder assignResponsiveSequencerAsPrimary(@NonNull Set<String> endpoints) {
-
-        List<String> modifiedSequencerServers = new ArrayList<>(layout.getSequencers());
-        for (int i = 0; i < modifiedSequencerServers.size(); i++) {
-            String sequencerServer = modifiedSequencerServers.get(i);
-
-            // Add a sequencer server given sequencerServer is not already included in
-            // endpoints AND it is not included in the unresponsive nodes of the layout.
-            if (!endpoints.contains(sequencerServer) &&
-                !layout.getUnresponsiveServers().contains(sequencerServer)) {
-                modifiedSequencerServers.remove(sequencerServer);
-                modifiedSequencerServers.add(0, sequencerServer);
-                layout.setSequencers(modifiedSequencerServers);
-                return this;
+    public LayoutBuilder assignResponsiveSequencerAsPrimary(@NonNull Set<String> excludedNodes) {
+        List<String> modifiedSequencers = new LinkedList<>(layout.getSequencers());
+        String primarySequencer = null;
+        Set<String> logTails = layout.getLastSegment().getStripes().stream()
+                .map(LayoutStripe::getTailEndpoint)
+                .collect(Collectors.toSet());
+
+        // Elect a sequencer server as primary given that server is not already included
+        // in excludedNodes AND it is not included in the unresponsive nodes of the layout.
+        // Non log tail servers have high priority to be elected for better load distribution.
+        for (String sequencer : modifiedSequencers) {
+            if (excludedNodes.contains(sequencer) ||
+                    layout.getUnresponsiveServers().contains(sequencer)) {
+                continue;
+            }
+            if (primarySequencer == null) {
+                primarySequencer = sequencer;
+            }
+            if (!logTails.contains(sequencer)) {
+                primarySequencer = sequencer;
+                break;
             }
```

Comment:
A way to save if-clauses might be:
`primarySequencer = sequencer;
if (!logTails.contains(sequencer)) {
break;
}`
The only difference is that if there is no non-tail servers, your code picks the first sequencer in the loop iteration whereas mine picks the last one. Does this matter?

Reply:
Yea I actually thought about this, but I was afraid there are tests relying on selecting the first node, so I kept it same as before. Looks like it's not the case, just did the change, thanks.